File: src/mysequator/engine/calibration.py
from __future__ import annotations

from pathlib import Path

import numpy as np

from .io import load_image


def median_frame(paths: list[Path]) -> np.ndarray | None:
    if not paths:
        return None
    frames = [load_image(path) for path in paths]
    _assert_same_shape(frames)
    return np.median(np.stack(frames, axis=0), axis=0).astype(np.float32)


def calibrate(image: np.ndarray, dark: np.ndarray | None, flat: np.ndarray | None) -> np.ndarray:
    out = image.astype(np.float32, copy=True)
    if dark is not None:
        _assert_shape(out, dark)
        out = np.maximum(out - dark, 0.0)

    if flat is not None:
        _assert_shape(out, flat)
        flat_safe = np.maximum(flat, 1e-4)
        channel_median = np.median(flat_safe.reshape(-1, flat_safe.shape[-1]), axis=0)
        out = out * channel_median.reshape(1, 1, -1) / flat_safe

    return np.clip(out, 0.0, 1.0)


def _assert_same_shape(frames: list[np.ndarray]) -> None:
    first_shape = frames[0].shape
    for frame in frames[1:]:
        _assert_shape(frames[0], frame)
    if len(first_shape) != 3 or first_shape[2] != 3:
        raise ValueError("Images must be RGB.")


def _assert_shape(a: np.ndarray, b: np.ndarray) -> None:
    if a.shape != b.shape:
        raise ValueError(f"Image shape mismatch: {a.shape} != {b.shape}")

File: Sources/MySequatorCore/Calibration.swift
import Foundation

func medianFrame(_ paths: [URL]) throws -> FloatRGBImage? {
    guard !paths.isEmpty else { return nil }
    let frames = try paths.map(loadImage)
    try assertSameShape(frames)

    var output = FloatRGBImage(width: frames[0].width, height: frames[0].height)
    var values = Array(repeating: Float(0), count: frames.count)
    for index in output.data.indices {
        for frameIndex in frames.indices {
            values[frameIndex] = frames[frameIndex].data[index]
        }
        output.data[index] = values.median()
    }
    return output
}

func calibrate(_ image: FloatRGBImage, dark: FloatRGBImage?, flat: FloatRGBImage?) throws -> FloatRGBImage {
    var out = image
    if let dark {
        guard image.hasSameShape(as: dark) else {
            throw MySequatorError.shapeMismatch("Dark frame shape does not match image.")
        }
        for i in out.data.indices {
            out.data[i] = max(out.data[i] - dark.data[i], 0)
        }
    }

    if let flat {
        guard image.hasSameShape(as: flat) else {
            throw MySequatorError.shapeMismatch("Flat frame shape does not match image.")
        }
        var medians = [Float](repeating: 0, count: 3)
        for channel in 0..<3 {
            var values = [Float]()
            values.reserveCapacity(flat.pixelCount)
            for pixel in 0..<flat.pixelCount {
                values.append(max(flat.data[pixel * 3 + channel], 0.0001))
            }
            medians[channel] = values.median()
        }
        for pixel in 0..<out.pixelCount {
            for channel in 0..<3 {
                let idx = pixel * 3 + channel
                let safeFlat = max(flat.data[idx], 0.0001)
                out.data[idx] = out.data[idx] * medians[channel] / safeFlat
            }
        }
    }

    return out.clipped()
}

func assertSameShape(_ frames: [FloatRGBImage]) throws {
    guard let first = frames.first else { return }
    for frame in frames.dropFirst() where !frame.hasSameShape(as: first) {
        throw MySequatorError.shapeMismatch("Images must have the same dimensions.")
    }
}
